File: problems/Find All Anagrams in a String/findAnagrams.cs
public class Solution {
    public IList<int> FindAnagrams(string s, string p) {
        var result = new List<int>();
        var sLen = s.Length;
        var pLen = p.Length;

        if (sLen < pLen) {
            return result;
        }

        var pLetterFrequencies = new int[26];
        var sLetterFrequencies = new int[26];

        foreach (var letter in p) {
            ++pLetterFrequencies[letter - 'a'];
        }

        for (var i = 0; sLen > i; ++i) {
            ++sLetterFrequencies[s[i] - 'a'];

            if (pLen <= i) {
                --sLetterFrequencies[s[i - pLen] - 'a'];
            }

            if (pLetterFrequencies.SequenceEqual(sLetterFrequencies)) {
                result.Add(1 + i - pLen);
            }
        }

        return result;
    }
}

File: problems/Find All Anagrams in a String/findAnagrams.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> result = new ArrayList<>();
        int[] letterFrequencies = new int[26];
        int letterCounter = 0;
        int pSize = p.length();

        for (char letter : p.toCharArray()) {
            ++letterFrequencies[letter - 'a'];
        }

        for (int idx = 0; s.length() > idx; ++idx) {
            if (0 < letterFrequencies[s.charAt(idx) - 'a']--) {
                ++letterCounter;
            }

            if (pSize <= idx && 0 < ++letterFrequencies[s.charAt(idx - pSize) - 'a']) {
                --letterCounter;
            }

            if (pSize == letterCounter) {
                result.add(1 + idx - pSize);
            }
        }

        return result;
    }
}
